File: src/main/java/org/frawa/elmtest/run/ElmTestRunConfigurationProducer.java
package org.frawa.elmtest.run;

import com.intellij.execution.Location;
import com.intellij.execution.actions.ConfigurationContext;
import com.intellij.execution.actions.LazyRunConfigurationProducer;
import com.intellij.execution.configurations.ConfigurationFactory;
import com.intellij.openapi.components.ServiceManager;
import com.intellij.openapi.util.Ref;
import com.intellij.psi.PsiElement;
import org.elm.workspace.ElmWorkspaceService;
import org.frawa.elmtest.core.ElmProjectTestsHelper;
import org.jetbrains.annotations.NotNull;

import java.nio.file.Path;
import java.util.Optional;

public class ElmTestRunConfigurationProducer extends LazyRunConfigurationProducer<ElmTestRunConfiguration> {

    @NotNull
    @Override
    public ConfigurationFactory getConfigurationFactory() {
        return new ElmTestConfigurationFactory(new ElmTestRunConfigurationType());
    }

    @Override
    protected boolean setupConfigurationFromContext(ElmTestRunConfiguration configuration, ConfigurationContext context, Ref<PsiElement> sourceElement) {
        return getCandidateElmFolder(context)
                .map(folder -> {
                    configuration.options.elmFolder = folder;
                    configuration.setGeneratedName();
                    return true;
                })
                .orElse(false);
    }

    @Override
    public boolean isConfigurationFromContext(ElmTestRunConfiguration configuration, ConfigurationContext context) {
        return getCandidateElmFolder(context)
                .map(folder -> folder.equals(configuration.options.elmFolder))
                .orElse(false);
    }

    private Optional<String> getCandidateElmFolder(ConfigurationContext context) {
        if (context == null) return Optional.empty();

        ElmWorkspaceService elmWorkspace = ServiceManager.getService(
                context.getProject(), ElmWorkspaceService.class);

        return Optional.of(context)
                .map(ConfigurationContext::getLocation)
                .map(Location::getVirtualFile)
                .map(elmWorkspace::findProjectForFile)
                .flatMap(Optional::ofNullable)
                .map(ElmProjectTestsHelper::elmFolderForTesting)
                .map(Path::toString);
    }
}

File: src/main/kotlin/org/elm/ide/test/run/ElmTestRunConfigurationProducer.kt
package org.elm.ide.test.run

import com.intellij.execution.actions.ConfigurationContext
import com.intellij.execution.actions.LazyRunConfigurationProducer
import com.intellij.openapi.util.Ref
import com.intellij.psi.PsiElement
import org.elm.workspace.elmWorkspace

class ElmTestRunConfigurationProducer : LazyRunConfigurationProducer<ElmTestRunConfiguration>() {

    override fun getConfigurationFactory() =
            ElmTestConfigurationFactory(ElmTestRunConfigurationType())

    override fun setupConfigurationFromContext(configuration: ElmTestRunConfiguration, context: ConfigurationContext, sourceElement: Ref<PsiElement>): Boolean {
        val elmFolder = getCandidateElmFolder(context) ?: return false
        configuration.options.elmFolder = elmFolder
        configuration.setGeneratedName()
        return true
    }

    override fun isConfigurationFromContext(configuration: ElmTestRunConfiguration, context: ConfigurationContext): Boolean {
        val elmFolder = getCandidateElmFolder(context) ?: return false
        return elmFolder == configuration.options.elmFolder
    }

    private fun getCandidateElmFolder(context: ConfigurationContext): String? {
        val vfile = context.location?.virtualFile ?: return null
        val elmProject = context.project.elmWorkspace.findProjectForFile(vfile) ?: return null
        return elmProject.projectDirPath.toString()
    }
}
